src/compare-tree/query.js

```diff
@@ -0,0 +1,37 @@
+import Emitter from '../emitter'
+
+import {
+  get,
+  inherit
+} from '../utils'
+
+import {
+  getKeyPaths
+} from './key-path'
+
+export default function Query (id, keys) {
+  Emitter.call(this)
+
+  this.revision = -1
+  this.id = id
+  this.keys = getKeyPaths(keys)
+}
+
+inherit(Query, Emitter, {
+  extract (state) {
+    let len = this.keys.length
+    let values = Array(len)
+
+    for (var i = 0; i < len; i++) {
+      values[i] = get(state, this.keys[i])
+    }
```

Comment:
It just isn't nearly as fast. It allocates memory for the anonymous function, and has to pay the penalty to maintain the JS closure.

But I could run the numbers. It might not really matter.